### PR TITLE
RATIS-1525. Add step down leader proto and client related code

### PR DESCRIPTION
**Pull request by codings-dan**
## What changes were proposed in this pull request?

 subtask to support safemode: add step down leader proto and client related code

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/RATIS-1525

## How was this patch tested?
UT

### CODE REVIEW

File: ratis-client/src/main/java/org/apache/ratis/client/api/AdminApi.java [hunk 1]
```diff
@@ -40,4 +40,7 @@ default RaftClientReply setConfiguration(RaftPeer[] serversInNewConf) throws IOE
 
   /** Transfer leadership to the given server.*/
   RaftClientReply transferLeadership(RaftPeerId newLeader, long timeoutMs) throws IOException;
+
+  /** Step down the leader.*/
+  RaftClientReply stepDownLeader(RaftPeerId newLeader, long timeoutMs) throws IOException;
```
Comment (szetszwo) [inline ratis-client/src/main/java/org/apache/ratis/client/api/AdminApi.java:45]: Let's reuse transferLeadership with newLeader==null.

**Comment by codings-dan** [comment]
@szetszwo PTAL again, thx!

File: ratis-client/src/main/java/org/apache/ratis/client/impl/ClientProtoUtils.java [hunk 2]
```diff
@@ -519,22 +520,38 @@ static SetConfigurationRequestProto toSetConfigurationRequestProto(
...
-    final RaftPeer newLeader = ProtoUtils.toRaftPeer(p.getNewLeader());
-    return new TransferLeadershipRequest(
-        ClientId.valueOf(m.getRequestorId()),
-        RaftPeerId.valueOf(m.getReplyId()),
-        ProtoUtils.toRaftGroupId(m.getRaftGroupId()),
-        p.getRpcRequest().getCallId(),
-        newLeader.getId(),
-        m.getTimeoutMs());
+    String address = p.getNewLeaderOrBuilder().getAddress();
+    if (!Objects.equals(address, "")) {
```
[the remark this answers is not in the document]
Comment (codings-dan) [reply ratis-client/src/main/java/org/apache/ratis/client/impl/ClientProtoUtils.java]: done

File: ratis-client/src/main/java/org/apache/ratis/client/impl/ClientProtoUtils.java [hunk 3]
```diff
@@ -519,22 +520,38 @@ static SetConfigurationRequestProto toSetConfigurationRequestProto(
...
+    if (request.getNewLeader() != null) {
+      return TransferLeadershipRequestProto.newBuilder()
+          .setRpcRequest(toRaftRpcRequestProtoBuilder(request))
+          .setNewLeader(RaftPeer.newBuilder().setId(request.getNewLeader()).build().getRaftPeerProto())
+          .build();
+    } else {
+      return TransferLeadershipRequestProto.newBuilder()
+          .setRpcRequest(toRaftRpcRequestProtoBuilder(request))
+          .build();
+    }
```
[the remark this answers is not in the document]
Comment (codings-dan) [reply ratis-client/src/main/java/org/apache/ratis/client/impl/ClientProtoUtils.java]: done

File: ratis-client/src/main/java/org/apache/ratis/client/impl/ClientProtoUtils.java [hunk 4]
```diff
@@ -18,6 +18,7 @@
 package org.apache.ratis.client.impl;
 
 import java.nio.ByteBuffer;
+import java.util.Objects;
```
Comment (szetszwo) [inline ratis-client/src/main/java/org/apache/ratis/client/impl/ClientProtoUtils.java:21]: There is a checkstyle warning
[235](https://github.com/apache/ratis/runs/5218732837?check_suite_focus=true#step:4:235)
```
Error:  src/main/java/org/apache/ratis/client/impl/ClientProtoUtils.java:[21,8] (imports) UnusedImports: Unused import - java.util.Objects.
```

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**